module/zfs/dsl_scan.c

```diff
@@ -4287,35 +4290,49 @@ dsl_scan_sync(dsl_pool_t *dp, dmu_tx_t *tx)
 	dsl_scan_t *scn = dp->dp_scan;
 	spa_t *spa = dp->dp_spa;
 	state_sync_type_t sync_type = SYNC_OPTIONAL;
+	uint64_t to_issue, issued;
+	int restart_early;
 
 	if (spa->spa_resilver_deferred &&
 	    !spa_feature_is_active(dp->dp_spa, SPA_FEATURE_RESILVER_DEFER))
 		spa_feature_incr(spa, SPA_FEATURE_RESILVER_DEFER, tx);
 
+	/*
+	 * Only process scans in sync pass 1.
+	 */
+	if (spa_sync_pass(spa) > 1)
+		return;
+
+	/*
+	 * issued/to_issue as presented to the user
+	 * in print_scan_scrub_resilver_status() issued/total_i
+	 * @ cmd/zpool/zpool_main.c
+	 */
+	to_issue = scn->scn_phys.scn_to_examine - scn->scn_phys.scn_skipped;
+	issued = scn->scn_issued_before_pass + spa->spa_scan_pass_issued;
+	restart_early = spa->spa_resilver_deferred && (
+	    zfs_resilver_disable_defer ||
+	    (issued < (to_issue * zfs_resilver_defer_percent / 100)));
+
 	/*
 	 * Check for scn_restart_txg before checking spa_load_state, so
 	 * that we can restart an old-style scan while the pool is being
 	 * imported (see dsl_scan_init). We also restart scans if there
 	 * is a deferred resilver and the user has manually disabled
-	 * deferred resilvers via the tunable.
+	 * deferred resilvers via zfs_resilver_disable_defer, or if the
+	 * curent scan progress is below zfs_resilver_defer_percent.
 	 */
-	if (dsl_scan_restarting(scn, tx) ||
-	    (spa->spa_resilver_deferred && zfs_resilver_disable_defer)) {
+	if (dsl_scan_restarting(scn, tx) || restart_early) {
 		pool_scan_func_t func = POOL_SCAN_SCRUB;
 		dsl_scan_done(scn, B_FALSE, tx);
 		if (vdev_resilver_needed(spa->spa_root_vdev, NULL, NULL))
 			func = POOL_SCAN_RESILVER;
-		zfs_dbgmsg("restarting scan func=%u on %s txg=%llu",
-		    func, dp->dp_spa->spa_name, (longlong_t)tx->tx_txg);
+		zfs_dbgmsg("restarting scan func=%u on %s txg=%llu early=%d",
+		    func, dp->dp_spa->spa_name, (longlong_t)tx->tx_txg,
+		    restart_early);
 		dsl_scan_setup_sync(&func, tx);
 	}
 
-	/*
-	 * Only process scans in sync pass 1.
-	 */
-	if (spa_sync_pass(spa) > 1)
-		return;
```

Comment:
Was this moved up solely as an optimization?

Reply:
AFAIK if I had to move it b/c it would keep restarting the scan in an endless loop due to the second part of `(dsl_scan_restarting(scn, tx) || restart_early)` condition, which I reused to put the scan in restarting state, so originally this condition would hit only once, but after my change the `restart_early` would keep hitting; this early restart logic should be only evaluated once per whole scan - if you have an idea how to handle this better, I'll gladly rework it :)

Reply:
Ah, that makes sense.  Moving the check up is fine.